DRILL-6179: Added pcapng-format support

Vlad-Storona opened the pull request:
See DRILL-6179 for details.
https://issues.apache.org/jira/browse/DRILL-6179

kkhatua commented:
@amansinha100  / @parthchandra / @arina-ielchiieva is this ready for a review?

arina-ielchiieva commented:
@kkhatua better ping @tdunning,

arina-ielchiieva commented:
@Vlad-Storona could you please rebase to the latest master and confirm that PR is ready for review?

Vlad-Storona commented:
@arina-ielchiieva I rebased project to the latest master, it is ready for review.

arina-ielchiieva commented:
@parthchandra / @paul-rogers since you have reviewed PR about pcap, could you please review this one as well?

priteshm commented:
@Vlad-Storona will you get a chance to address the code review comments in the next few days (in time for it to be considered for the 1.14 release)?

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngFormatPlugin.java:
```diff
@@ -0,0 +1,111 @@
+  public PcapngFormatPlugin(String name, DrillbitContext context, Configuration fsConf, StoragePluginConfig config, PcapngFormatConfig formatPluginConfig) {
+    super(name, context, fsConf, config, formatPluginConfig, true,
+        false, true, false,
+        Lists.newArrayList("pcapng"), "pcapng");
```
Yes, this suffix is a common standard for these files. But I suppose that I must specify it in the PcapngFormatConfig as in other plugins.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngFormatPlugin.java:
```diff
@@ -0,0 +1,111 @@
+  private static class PcapngFormatMatcher extends BasicFormatMatcher {
+
+    public PcapngFormatMatcher(PcapngFormatPlugin plugin) {
+      super(plugin, ImmutableList.of(Pattern.compile(".*\\.pcapng$")), ImmutableList.of());
```
Yes, I totally agree with you, this class is unnecessary for this plugin.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+
+  @Override
+  public void close() throws Exception {
+    in.close();
```
Ok, I will add this verification.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+                            final List<SchemaPath> projectedColumns) {
+    this.fs = fileSystem;
+    this.pathToFile = fs.makeQualified(new Path(pathToFile));
+    this.projectedColumns = projectedColumns;
```
Yes, it is my mistake, I will move ```setColumns(columns)``` to the constructor as in other plugins.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  private ImmutableList<ProjectedColumnInfo> getProjectedColsIfItNull() {
+    return projectedCols != null ? projectedCols : initCols(new Schema());
```
This plugin supports all these cases, but as I understand in a wrong way. I will rewrite it.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+  private ImmutableList<ProjectedColumnInfo> initCols(final Schema schema) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+
+    for (String columnName : schema.getColumns().keySet()) {
```
Sure, we need to build a project list in the order defined by the projected column list, not by a hard-coded order as now. I will change it. Also, thanks for the explanation about the sensitivity of a drill and about expected behaviour with missing columns. I will solve these problems and will add tests for these cases.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+  private ImmutableList<ProjectedColumnInfo> initCols(final Schema schema) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+
+    for (String columnName : schema.getColumns().keySet()) {
```
This code supports 'star' query, but with column order that defined in the ```Schema``` class. As I understand from previous comments we need to arrange the columns based on the projected list. Because of that, I will change this behaviour.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+      final Class<? extends ValueVector> clazz = TypeHelper.getValueVectorClass(
+          minorType, majorType.getMode());
+      ValueVector vector = output.addField(field, clazz);
+      vector.allocateNew();
```
Ok, thanks.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+      return vector;
+
+    } catch (SchemaChangeException sce) {
+      throw new IllegalStateException("The addition of this field is incompatible with this OutputMutator's capabilities");
```
Sure.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+
+  private void addBasicInfoToTable(final IEnhancedPacketBLock block, final int count) {
+    for (ProjectedColumnInfo pci : projectedCols) {
+      switch (pci.columnName) {
```
Okay, I will try to do something with it for solving this problem. Also, thanks for the example.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/Schema.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.Map;
+
+public class Schema {
+  private final static Map<String, PcapngTypes> columns = new LinkedHashMap<>();
```
Yes, this is ordered map class, because of this I have a hard-coded order of columns. I will change it because we must arrange the order of columns based on a projection, not on a hard-coded order, now I understand it.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,260 @@
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+
+    for (String columnName : schema.getColumns().keySet()) {
+      final PcapngTypes type = schema.getColumns().get(columnName);
```
Ok, I will solve it.

Vlad-Storona commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/pcapng/TestPcapngRecordReader.java:
```diff
@@ -0,0 +1,68 @@
+  @Test
+  public void testCountQuery() throws Exception {
+    Assert.assertEquals(1, testSql("select count(*) from dfs.`store/pcapng/sniff.pcapng`"));
+    Assert.assertEquals(1, testSql("select count(*) from dfs.`store/pcapng/example.pcapng`"));
```
I will rewrite this test because this it doesn't check what result this query will return. And I have a question. How can I write a test to check how many vectors are allocated for a query?

Vlad-Storona commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/pcapng/TestPcapngRecordReader.java:
```diff
@@ -0,0 +1,68 @@
+    String query = "EXPLAIN PLAN for select * from dfs.`store/pcapng/sniff.pcapng`";
+    String plan = getPlanInString(query, JSON_FORMAT);
+    Assert.assertEquals(123, testPhysical(plan));
+  }
```
Okay, I will add all these cases, thanks. Also, thanks for the example.

paul-rogers replied:
Looks good, thanks.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/decoder/structure/IEnhancedPacketBLock.java:
```diff
@@ -0,0 +1,63 @@
+ * limitations under the License.
+ */
+package org.apache.drill.exec.store.pcapng.decoder.structure;
+
```
It is my mistake, I didn't need to copy this code from the library. It was a wrong decision. Because of this, I will add library from the maven.

Vlad-Storona commented:
@paul-rogers thanks for the great review. I tried to solve all the issues described above. If I missed something, please let me know. Also, can you take a look at new changes?

@priteshm I will try to make changes in time.

arina-ielchiieva commented:
@paul-rogers could you please do final code review? Vlad has addressed code review comments and we plan to include this change in 1.14.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+  private ImmutableList<ProjectedColumnInfo> projectCols(final List<SchemaPath> columns) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    for (SchemaPath schemaPath : columns) {
+      String projectedName = schemaPath.rootName();
```
As it turns out, Drill lets us refer to columns in the project list in a variety of forms:

```
SELECT a, b[3], c.d ...
```

The code here uses just the root name. This works fine for the `a` case. But, it is incorrect for the `b[3]` and `c.d` cases. We need to check that the `schemaPath` has no children. If it does have children, then treat the column as undefined: give it a nullable int vector.

That is, there are no arrays in pcapng (as I understand it), so `b[3]` can't refer to a valid element of an array. In Drill, such an undefined column is replaced by a nullable int.

Vlad-Storona replied:
Sure, I will add this cheks. Also, I will add tests for this cases.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    for (SchemaPath schemaPath : columns) {
+      String projectedName = schemaPath.rootName();
+      if (Schema.getColumns().containsKey(projectedName.toLowerCase())) {
```
This code assumes that all columns are matched. Actually, if the user requests the `foo` column which does not exist in the pcapng file, the proper behavior is, as described above, to create a dummy nullable int column, with the requested name.

Said another way, if the user asks for three columns, this reader must provide three columns, even if it has to "make up" nullable int columns to do so.

Looks like the code intended to handle this case: there is a `DummyImpl` that seems like what is needed here if there is no match for the column name.

Vlad-Storona replied:
This works because Drill sets to all of the columns a MajorType as optional before initialization in a reader.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/schema/Schema.java:
```diff
@@ -0,0 +1,90 @@
+  private final static Map<String, Column> columns = new HashMap<>();
+
+  static {
+    columns.put("timestamp", new TimestampImpl());
```
Minor style issue. You can go ahead and include the `Impl` classes here as `static` nested classes. Puts all this tightly-coupled code in one easy place; makes it easier for readers.

Having each `Impl` in its own file is also fine, just a bit harder to read through many small files.

Vlad-Storona replied:
Sure, I'll move it to the one class.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+  @Override
+  public void close() throws Exception {
+    if (in != null) {
+      in.close();
```
`in = null;`?

Vlad-Storona replied:
Okay

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+  }
+
+  private ProjectedColumnInfo makeColumn(final String colName, final Column column) {
+    TypeProtos.MajorType majorType = Types.optional(column.getMinorType());
```
Can pcapng files omit fields? Or, if the user asks for `dst_ip`, say, will that field exist in every row?

If the value exists in every row, you'll get slightly better performance using a `required` type for that column. Doing so omits the nullable bit for each column.

Note that, for the `DummyImpl`, you do need the type to be `optional`. This might mean that the `MajorType` should be defined in the `Column` class so you can have some be required, others optional.

Vlad-Storona replied:
Yes, I missed it. Only two fields are required for every row and I will change its type to required. All other fields are optional for pcapng files.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/schema/Schema.java:
```diff
@@ -0,0 +1,90 @@
+import java.util.Map;
+import java.util.Set;
+
+public class Schema {
```
This `Schema` design is very nice: simple, easy to understand. Let's encourage others to use this kind of design for future readers.

Vlad-Storona replied:
Thanks.

paul-rogers commented on exec/java-exec/src/test/java/org/apache/drill/exec/store/pcapng/TestPcapngHeaders.java:
```diff
@@ -0,0 +1,154 @@
+
+    expectedSchema.add(MaterializedField.create("timestamp", Types.optional(TypeProtos.MinorType.TIMESTAMP)));
+    expectedSchema.add(MaterializedField.create("name", Types.optional(TypeProtos.MinorType.INT)));
+    expectedSchema.add(MaterializedField.create("color", Types.optional(TypeProtos.MinorType.INT)));
```
I wonder, on this test case, was it the pcapng code that created the Nullable Int, or somewhere else in Drill?

Vlad-Storona replied:
As I described above, Drill sets to all of the columns a MajorType as optional.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+
+  private ValueVector getValueVector(final TypeProtos.MinorType minorType,
+                                     final TypeProtos.MajorType majorType,
+                                     final MaterializedField field, final OutputMutator output) {
```
Thanks, I will simplify this.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java:
```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  private List<IEnhancedPacketBLock> getBatchOfBlocks() {
```
Ok, I will change it.

Vlad-Storona commented:
@paul-rogers thanks for the review. I solved all the issues. Also, I rebased to the latest master and solved conflicts. Can you take a look at changes? 

Also, I found a problem with a DRILL-6104, because of it I have had conflicts with a master branch. The problem was with proto files. In DRILL-6104 was edited generated files in `drill-protocol` module without regeneration. I updated the proto file with an information about my plugin and DRILL-6104.

arina-ielchiieva commented:
@cgivre FYI.

cgivre commented:
Thanks Arina and sorry about that.  For the future and for me to learn, what file should I have updated, or how should I have updated the protocols correctly?
Thanks,


> On Jul 19, 2018, at 08:17, Arina Ielchiieva <notifications@github.com> wrote:
> 
> @cgivre <https://github.com/cgivre> FYI.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/apache/drill/pull/1126#issuecomment-406256293>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AFQfvm88GBcfYjOuW0qPVUwlFWgBJkTrks5uIHjfgaJpZM4SPxHQ>.
>

arina-ielchiieva commented:
@cgivre that's easy :)
1. make changes only in .proto files (https://github.com/apache/drill/tree/master/protocol/src/main/protobuf).
2. regenerate them following the instruction (https://github.com/apache/drill/blob/master/protocol/readme.txt).
3. commit the changes.

arina-ielchiieva commented:
@cgivre I am not sure if Vlad's changes make into the release. Maybe you should go ahead and create Jira and PR with protobuf fix?

arina-ielchiieva commented:
@vdiravka created Jira https://issues.apache.org/jira/browse/DRILL-6627

vdiravka commented:
@Vlad-Storona @cgivre @arina-ielchiieva I have created ticket to add the missed operator to the protobuf file: [DRILL-6627](https://issues.apache.org/jira/browse/DRILL-6627). PR for it: https://github.com/apache/drill/pull/1393

arina-ielchiieva commented:
@Vlad-Storona could you please rebase?
@paul-rogers could you please do final review? Unfortunately, this did not make it into 1.14 but I guess we should consider not missing 1.15.

Vlad-Storona commented:
@paul-rogers thanks for the great review.

arina-ielchiieva commented on protocol/src/main/protobuf/UserBitShared.proto:
```diff
@@ -343,6 +343,7 @@ enum CoreOperatorType {
   IMAGE_SUB_SCAN = 52;
   SEQUENCE_SUB_SCAN = 53;
   PARTITION_LIMIT = 54;
+  PCAPNG_SUB_SCAN = 55;
```
You also need to generate C++ code.

Vlad-Storona replied:
Sure.

Vlad-Storona commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngFormatPlugin.java:
```diff
@@ -0,0 +1,76 @@
+
+  @Override
+  public int getWriterOperatorType() {
+    return 0;
```
Okay, changed.

Vlad-Storona commented:
@arina-ielchiieva thanks for comments, I fixed issues and rebased to the latest master.

arina-ielchiieva commented:
+1
